events: Update subscriber list on peer_remove event for unsubscribed and never-subscribed streams.

sahil839 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
- Update subscriber list on peer_remove event for unsubscribed streams.
- Update subscriber list on peer_remove event for never subscribed streams.

 <!-- How have you tested? -->


 <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

timabbott commented on zerver/lib/events.py:
```diff
@@ -661,6 +661,10 @@ def name(sub: Dict[str, Any]) -> str:
+            for sub in state['unsubscribed']:
+                if (sub['stream_id'] == stream_id and
+                        user_id in sub['subscribers']):
+                    sub['subscribers'].remove(user_id)
```
It looks like the `peer_add` code path also ignores `unsubscribed`?  

In any case, givne that the code is the same in these inner loops, can we just do:

`for sub in state['subscribed'] + state['unsubscribed'] + state['never_subscribed']`?

sahil839 commented:
@timabbott Updated the PR with -
- Updating subscriber list for `unsubscribed` streams on `peer_add` event.
- Using `for sub in state['subscriptions'] + state['unsubscribed'] + state['never_subscribed']`.
- Added a refactoring commit to not include `realm_user` in `event_types`.
- Added commit to check `vacant` event dict.

timabbott commented:
@sahil839 can you update this following @showell's recent batch of PRs that removed the `vacate` event type?  

I think once that's done, this should be good to merge.  Sorry for all the delay with this one!

sahil839 commented:
@timabbott I have updated the PR with following changes -
- Removed the commit for checking vacate event.
- Added a commit for fixing a comment regarding vacate event. 
Apparently there is one more PR from Steve #16594 which will be conflicting with these changes.

timabbott commented:
@sahil839 great; I just merged Steve's PR; can your rebase?

sahil839 commented:
Rebased the PR.

timabbott commented:
Awesome, merged, thanks @sahil839!
